R/splatPop-estimate.R

```diff
@@ -0,0 +1,176 @@
+                        eqtl.ES.rate = unname(fit$estimate["rate"]))
+
+    return(params)
+}
+
+#' Estimate gene mean and gene mean variance parameters
+#'
+#' The Shapiro-Wilks test is used to determine if the library sizes are
+#' normally distributed. If so a normal distribution is fitted to the library
+#' sizes, if not (most cases) a log-normal distribution is fitted and the
+#' estimated parameters are added to the params object. See
+#' \code{\link[fitdistrplus]{fitdist}} for details on the fitting.
+#'
+#' @param params splatPopParams object containing parameters for the 
+#'         simulation of the mean expression levels for the population.
+#'        See \code{\link{splatPopParams}} for details.
+#' @param means Dataframe of real gene means across a population, where 
+#'        each row is a gene and each column is an individual in the population.
+#'
+#' @details
+#' Parameters for the mean gamma distribution are estimated by fitting the mean
+#' (across the population) expression of genes that meet the criteria (<50% of
+#' samples have exp <0.1) and parameters for the cv gamma distribution are 
+#' estimated for each bin of mean expression using the cv of expression across
+#' the population for genes in that bin. Both are fit using 
+#' \code{\link[fitdistrplus]{fitdist}}. The "Nelder-Mead" method is used to fit
+#' the mean gamma distribution and the maximum goodness-of-fit estimation 
+#' method is used to minimise the Cramer-von Mises distance for the CV 
+#' distribution.
+#' 
+#' @return params object with estimated values.
+#' @importFrom stats quantile
+#' @importFrom grDevices boxplot.stats
+#' @importFrom matrixStats rowMedians
+#' 
+splatPopEstimateMeanCV <- function(params, means) {
+    
+    # Test input gene means
+    if ((anyNA(means) | !(validObject(rowSums(means))))) {
+        stop("Incorrect format or NAs present in gene.means. See example data.")
+    }
+    
+    # Remove genes with low variance/low means
+    abv_thr <- data.frame(perc = (rowSums(means >= 0.1)/ncol(means)),
+                          gene_id = row.names(means))
+    
+    genes <- row.names(abv_thr[abv_thr$perc > 0.5, ])
+    means <- means[genes, ]
+
+    # Calculate mean expression parameters
+    row_means <- rowMedians(as.matrix(means))
+    names(row_means) <- row.names(means)
+    mfit <- fitdistrplus::fitdist(row_means, "gamma",
+                                  optim.method="Nelder-Mead")
+    
+    # Calculate CV parameters for genes based on 10 expresion mean bins
+    nbins <- getParam(params, "pop.cv.bins")
+    bins <- split(row_means, cut(row_means, quantile(row_means,(0:nbins)/nbins), 
+                             include.lowest=TRUE))
+    cvparams <- data.frame(start = character(), end = character(),
+                           shape = character(), rate = character(), 
+                           stringsAsFactors=FALSE)
+
+    for(b in names(bins)){
+        stst <- unlist(strsplit(gsub("\\)|\\(|\\[|\\]", "", b), ","))
```

Comment:
I changed the variable name from stst (start_stop) to min_max to make it more clear what that line is doing. I also made a new variable called re.brack_paren and set that equal to "\\[|\\]|\\)|\\(" so that it is clear that gsub call is just removing special characters.